substreams-helper/src/storage.rs

```diff
@@ -34,12 +124,317 @@ impl Uint256 {
+        panic!("Field {} not found", name);
+    }
+}
+
+impl StorageLayout for Struct {
+    fn size(&self) -> usize {
+        size_to_slots(self.cumulative_size) * SLOT_SIZE
+    }
+
+    fn decode(
+        &mut self,
+        slots: Vec<Vec<u8>>,
+        _offset: Option<usize>,
+    ) -> Result<(), StorageDecodingError> {
+        if slots.len() < self.current_slot {
+            return Err(StorageDecodingError::new(format!(
+                "Invalid number of slots for Struct, expected {}, got {}",
+                size_to_slots(self.size()),
+                slots.len()
+            )));
+        }
+
+        let field_count = self.fields.len();
+        for i in 0..field_count {
+            let field = &mut self.fields[i];
+            field.value.decode(
+                slots[field.start_slot as usize..field.end_slot as usize].to_vec(),
+                Some(field.offset),
+            )?;
+        }
+        Ok(())
+    }
+
+    fn set_slot(&mut self, slot: BigInt) {
+        self.slot = slot;
+    }
+}
+
+pub struct Array<T: StorageLayout> {
+    pub slot: BigInt,
+    pub item: T,
+}
+
+impl<T: StorageLayout> Array<T> {
+    pub fn new(slot: BigInt, item: T) -> Self {
+        let mut arr = Array {
+            slot: slot,
+            item: item,
+        };
+
+        let item_slot = keccak256(arr.encoded_slot());
+        arr.item.set_slot(BigInt::abi_decode(item_slot).unwrap());
+        arr
+    }
+
+    /// Will return a list of all storage changes involving this array, from a given list of changes.
+    pub fn filter_array_changes(
+        &self,
+        changes: Vec<StorageChange>,
+        expected_array_len: BigInt,
+    ) -> Vec<StorageChange> {
+        let mut filtered_changes = vec![];
+        for change in changes {
+            let index = self.infer_array_index_from_storage_key(change.change.key.clone());
+            if index.is_none() {
+                continue;
+            }
+
+            if index.unwrap() <= expected_array_len {
+                filtered_changes.push(change);
+            }
+        }
+        filtered_changes
+    }
+
+    pub fn infer_array_index_from_storage_key(&self, key: Vec<u8>) -> Option<BigInt> {
+        let first_elem = self.storage_key_at_index(BigInt::zero());
+        let target_val = BigInt::abi_decode(key).unwrap();
+        let first_val = BigInt::abi_decode(first_elem).unwrap();
+
+        if target_val < first_val {
+            return None;
+        }
+
+        let slot_diff = target_val - first_val;
+        let slots_per_item = BigInt::from(size_to_slots(self.item.size()));
+        if slot_diff.modulo(slots_per_item.as_ref()) != BigInt::zero() {
+            return None;
+        }
+
+        Some(slot_diff / slots_per_item)
+    }
+
+    /// Returns the EVM storage key where the value of a given array index is stored.
+    pub fn storage_key_at_index(&self, index: BigInt) -> Vec<u8> {
+        let slots_per_item = BigInt::from(size_to_slots(self.item.size()));
```

Comment:
size_to_slots here is usize. I think it would be fine to write the modulo logic to accept a usize arg instead and return a usize arg. Will save on the BigInt wrapping :)

Reply:
I'll still need to wrap the usize in a BigInt, since I need to divide a BigInt by the usize in the calculation of the modulo. It will just be done lower down the stack.

Changed it, the implementation of the modulo now looks something like:
```
impl<T> NumberModulo<T> for BigInt
where
    T: Into<BigInt> + Clone,
{
    fn modulo(&self, other: T) -> Self {
        let division = self.to_owned() / other.clone();
        self.to_owned() - division * other
    }
}
```

Types implementing `Into<BigInt>` are already usable in divisions and multiplications.